# Tutorial 1: Project Setup and Data Understanding

## Instacart Market Basket Analysis Dataset

### Advanced Data Analytics (BI203)

In [12]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [13]:
orders = pd.read_csv("instacart/orders.csv")
products = pd.read_csv("instacart/products.csv")
order_products = pd.read_csv("instacart/order_products__prior.csv")

In [18]:
orders.head()

,order_id,user_id,eval_set,order_number,order_dow,order_hour_of_day,days_since_prior_order
0,2539329,1,prior,1,2,8,NaN
1,2398795,1,prior,2,3,7,15.0
2,473747,1,prior,3,3,12,21.0
3,2254736,1,prior,4,4,7,29.0
4,431534,1,prior,5,4,15,28.0


In [19]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(5), object(1)
memory usage: 182.7+ MB


In [20]:
orders.shape

(3421083, 7)

In [24]:
orders.isna().sum().sort_values(ascending=False)

,0
days_since_prior_order,206209
user_id,0
order_id,0
eval_set,0
order_number,0
order_dow,0
order_hour_of_day,0


## Observations for Orders Table

- The orders dataset contains 3421083 rows and 7 columns.

- The order_id column appears to be the primary identifier.

- The user_id column identifies customers.

- Missing values appear in the days_since_prior_order column.

- The dataset contains customer order history information.

In [21]:
products.head()


,product_id,product_name,aisle_id,department_id
0,1,Chocolate Sandwich Cookies,61,19
1,2,All-Seasons Salt,104,13
2,3,Robust Golden Unsweetened Oolong Tea,94,7
3,4,Smart Ones Classic Favorites Mini Rigatoni Wit...,38,1
4,5,Green Chile Anytime Sauce,5,13


In [22]:
products.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49688 entries, 0 to 49687
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   product_id     49688 non-null  int64 
 1   product_name   49688 non-null  object
 2   aisle_id       49688 non-null  int64 
 3   department_id  49688 non-null  int64 
dtypes: int64(3), object(1)
memory usage: 1.5+ MB


In [23]:
products.shape

(49688, 4)

In [25]:
products.isna().sum().sort_values(ascending=False)

,0
product_id,0
product_name,0
aisle_id,0
department_id,0


## Observations for Products Table

- The products dataset contains product information.
- product_id appears to be the primary key.
- aisle_id and department_id link products to other tables.
- The dataset contains product names and category information.

In [26]:
order_products.head()

,order_id,product_id,add_to_cart_order,reordered
0,2,33120,1,1
1,2,28985,2,1
2,2,9327,3,0
3,2,45918,4,1
4,2,30035,5,0


In [27]:
order_products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32434489 entries, 0 to 32434488
Data columns (total 4 columns):
 #   Column             Dtype
---  ------             -----
 0   order_id           int64
 1   product_id         int64
 2   add_to_cart_order  int64
 3   reordered          int64
dtypes: int64(4)
memory usage: 989.8 MB


In [28]:
order_products.shape

(32434489, 4)

In [29]:
order_products.isna().sum().sort_values(ascending=False)

,0
order_id,0
product_id,0
add_to_cart_order,0
reordered,0


## Observations for Order Products Table

- The dataset links products to customer orders.
- order_id connects this table to the orders table.
- product_id connects this table to the products table.
- reordered indicates whether a customer previously purchased a product.

# Data Dictionary

| Variable Name | Meaning | Data Type | Notes |
|---|---|---|---|
| order_id | Unique order identifier | Integer | Primary key |
| user_id | Customer identifier | Integer | Links customers to orders |
| order_number | Sequence number of customer orders | Integer | Shows purchase order |
| order_dow | Day of the week order was placed | Integer | 0–6 |
| order_hour_of_day | Hour order was placed | Integer | 0–23 |
| days_since_prior_order | Days since previous order | Float | Contains missing values |
| product_id | Unique product identifier | Integer | Primary key |
| product_name | Name of product | String | Product description |
| aisle_id | Aisle identifier | Integer | Links to aisles |
| department_id | Department identifier | Integer | Links to departments |

# Table Relationships

The Instacart dataset is relational and consists of multiple linked tables.

## Orders Table
- order_id is the primary key.
- user_id identifies customers.

## Products Table
- product_id is the primary key.
- aisle_id and department_id classify products.

## Order Products Table
- order_id links to the orders table.
- product_id links to the products table.

The order_products table acts as a bridge table connecting customer orders with purchased products.

# Business Understanding

The Instacart Market Basket Analysis dataset represents an online grocery retail business where customers place orders for various products. The dataset captures customer purchasing behaviour, product information, and relationships between orders and items purchased.

This data can help businesses understand customer buying patterns and improve operational decision-making. For example, analysts can identify which products are frequently purchased together, which products are reordered most often, and which times of day experience the highest order activity.

The dataset can support several important business decisions. Marketing teams can use the data to create personalised product recommendations and targeted promotions. Inventory managers can use purchasing trends to improve stock planning and reduce shortages. Management can also use customer behaviour analysis to improve customer retention and increase sales performance.

From an analytics perspective, the dataset supports descriptive, predictive, and prescriptive analysis. Descriptive analytics can summarise customer ordering behaviour and popular products. Predictive analytics can estimate future purchasing behaviour or predict product reordering patterns. Prescriptive analytics can recommend business actions such as product placement strategies, recommendation systems, or promotional campaigns.

Overall, the dataset provides a realistic example of how large-scale business data can be analysed to support data-driven decision-making in retail organisations.

In [31]:
orders.to_csv("orders_interim.csv", index=False)

products.to_csv("products_interim.csv", index=False)

order_products.to_csv("order_products_interim.csv", index=False)

# Reflection

## 1. What type of business does this dataset represent?

This dataset represents an online grocery retail business where customers purchase products through repeated orders.

## 2. What decisions could this data support?

The data could support:
- product recommendation systems,
- inventory planning,
- customer behaviour analysis,
- marketing strategies, and
- demand forecasting.

## 3. Which tables are needed for descriptive, predictive, and prescriptive analysis?

- Descriptive analysis: orders, products, order_products
- Predictive analysis: orders and order_products
- Prescriptive analysis: all related tables together